### PR TITLE
Handle OSR from inlined synchronized method

### PR DESCRIPTION
**Pull request by cathyzhyi**
When a synchronized method exits through unhandled exception, it's
required by the specification to release the monitor before exits. The
VM unwind logic will handle this for synchronized methods that are not
inlined. For inlined ones, JIT inserts a synthetic handler to catch the
exception and do monitorExit and then rethrow the exception.

In case of OSR, the VM needs to be aware of those synthetic handlers to
resume execution correctly in the interpreter. Special catch type value
in the exception table entry is used for this purpose. Also, a new
helper for throwing exceptions from synthetic handlers is added.

This is the JIT side of the change.

see issue #9267

Signed-off-by: Yi Zhang <yizhang@ca.ibm.com>

### CODE REVIEW

**Comment by cathyzhyi** [comment]
this change depends on  https://github.com/eclipse/omr/pull/5127

**Comment by andrewcraik** [comment]
not sure what the travis vailure is about - I will continue the review

**Comment by andrewcraik** [comment]
Jenkins test sanity all jdk14

**Comment by cathyzhyi** [comment]
Seems the `TR_throwUnreportedException` defined in eclipse/omr#5127 is not picked up by the build yet.

**Comment by andrewcraik** [comment]
Jenkins test sanity all jdk14

**Comment by gacholio** [comment]
The 390 failure looks random:
```
11:52:35  FAILED: test_suspend
11:52:35  java.lang.AssertionError: Failed to resume thread
```

**Comment by gacholio** [comment]
Yep, the test is totally bogus:
```
			ct.resume();
			// Wait to be sure the resume has occurred.
			Thread.sleep(500);
			AssertJUnit.assertTrue("Failed to resume thread", orgval != t.getCheckVal());
```

File: runtime/compiler/runtime/MetaData.cpp [hunk 1]
```diff
@@ -158,6 +159,22 @@ createByteCodeInfoRange(
...
 
+/*
+ * use special catch type value to indicate to the VM the handler is synthetic handler so that OSR can be handled correctly
+ */
+static uint32_t
+useSpecialCatchTypeIfNecessary(TR_ExceptionTableEntry *e, bool fourByteOffsets, TR::Compilation *comp)
+   {
+   if (comp->getOption(TR_EnableOSR))
+      {
+      if (fourByteOffsets && e->_catchType == J9_CATCHTYPE_VALUE_FOR_THYNTHETIC_HANDLER_4BYTES ||
```
Comment (gacholio) [inline runtime/compiler/runtime/MetaData.cpp:170]: Some brackets would be good here (and perhaps essential - && and || likely have exactly the same precedence properties).

Also synthetic is misspelled.

Comment (andrewcraik) [reply runtime/compiler/runtime/MetaData.cpp:170]: @cathyzhyi please address above and I'll restart the build.

Comment (cathyzhyi) [reply runtime/compiler/runtime/MetaData.cpp]: Thanks for the comments. changed as suggested.

**Comment by gacholio** [comment]
jenkins test sanity zlinux jdk14

**Comment by gacholio** [comment]
Likely no need for full coverage again - just re-testing the one that failed (due to the bad test) before.

**Comment by gacholio** [comment]
@andrewcraik This ready for you to review/merge.

**Pull request merged**